core/src/com/unciv/logic/multiplayer/chat/ChatWebSocketManager.kt

```diff
@@ -0,0 +1,242 @@
+                url(chatUrl)
+                header(HttpHeaders.Authorization, UncivGame.Current.settings.multiplayer.getAuthHeader())
+            }
+
+            session!!.runCatching {
+                if (isActive) {
+                    if (errorReconnectionAttempts == 0) {
+                        println("ChatLog: Connected to WebSocket.")
+                        relayGlobalMessage("Successfully connected to WebSocket server!")
+                    } else if (errorReconnectionAttempts > 0) {
+                        println("ChatLog: Re-established webSocket connection.")
+                        relayGlobalMessage("Successfully re-established WebSocket connection!")
+                    }
+                    // we are successfully connected
+                    errorReconnectionAttempts = 0
+                }
+
+                val gameIds = ChatStore.getGameIds()
+                    .union(UncivGame.Current.onlineMultiplayer.games.mapNotNull { it.preview?.gameId })
+                this.sendSerialized(Message.Join(gameIds.toList()))
+
+                while (this.isActive) {
+                    val response = receiveDeserialized<Response>()
+                    when (response) {
+                        is Response.Chat -> EventBus.send(
+                            ChatMessageReceived(
+                                response.gameId, response.civName, response.message
+                            )
+                        )
+
+                        is Response.Error -> relayGlobalMessage("Error: [${response.message}]", "Server")
+                        is Response.JoinSuccess -> Unit
+                    }
+                    yield()
+                }
+            }
+                .onSuccess { restartSocket() }
+                .onFailure { handleWebSocketThrowables(it) }
+        } catch (e: Exception) {
+            handleWebSocketThrowables(e)
+        }
+    }
+
+    private fun startSocket() {
+        if (job?.isActive == true) return
+        job?.cancel()
+        job = Concurrency.run("MultiplayerChat") { startSession() }
+    }
+
+    /**
+     * By default, this gets autocancelled if the [job] is still running.
+     * Force mode will cancel the previous [job] and reassign a new one.
+     * This is helpfull when we need to reset [job] due to events.
+     */
+    @OptIn(DelicateCoroutinesApi::class)
+    private fun restartSocket(dueToError: Boolean = false, force: Boolean = false) {
+        if (dueToError) {
+            if (errorReconnectionAttempts++ > MAX_ERROR_RECONNECTION_ATTEMPTS) {
+                return
+            }
+        } else errorReconnectionAttempts = 0
+
+        GlobalScope.launch {
+            try {
+                delay(RECONNECT_TIME_MS)
+                if (job?.isActive == true && !force) return@launch
+
+                yield()
+                job?.cancel()
+                job = Concurrency.run("MultiplayerChat") { startSession() }
+            } catch (e: Exception) {
```

Comment:
Another catch-all. Which part of this function can **actually** throw an exception?

Reply:
I tried to check if any of the methods here can throw any exception and found this:
https://stackoverflow.com/questions/69189879/how-can-i-know-what-exceptions-could-be-throw-by-a-method-in-kotlin-android

In short there is no way to figure out if any methods in `Kotlin` throws any `Exception` at all.